[rollout] feat: Allow Custom AgentLoop

casper-hansen opened the pull request:
### What does this PR do?

The AgentLoop is a better implementation than ChatScheduler but it lacks extensibility. This PR adds the same custom import that you get with ChatScheduler to address the extensibility issue through `actor_rollout_ref.rollout.agent.custom_agent_loop`.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/volcengine/verl/pulls?q=is%3Apr+is%3Aopen+agent+loop
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ).

eric-haibin-lin commented on verl/experimental/agent_loop/agent_loop.py:
```diff
@@ -275,14 +276,24 @@ def get_agent_loop_class(self, agent_name: str) -> Type[AgentLoopBase]:
             return ToolAgentLoop
+
+        cfg_agent = getattr(
+            self.config.actor_rollout_ref.rollout.agent,
```
cfg_agent = self.config.actor_rollout_ref.rollout.agent.get("custom_agent_loop") should just work?

eric-haibin-lin replied:
could u add an unit test for the custom agent? thanks.
if you come across legacy config test error, you can ignore it. will be fixed by https://github.com/volcengine/verl/pull/2433

casper-hansen replied:
@eric-haibin-lin I'm pending a successful test run on the new unit test that I added, but is this an acceptable test in general? I just wanted to add a dummy test to simplify it as much as possible.

casper-hansen commented:
@eric-haibin-lin @wuxibin89 I addressed the code review and I have added a unit test to verify that the custom agent loop works. I have been able to run this locally with the test passing, showing that it works as intended.

```
pytest -s tests/experimental/agent_loop/test_basic_agent_loop.py -k test_custom_agent
```

casper-hansen commented:
@wuxibin89 merge conflicts resolved

eric-haibin-lin commented:
@casper-hansen there's some pre-commit lint error..
